File: 3147.taking-maximum-energy-from-the-mystic-dungeon/solution.go
func maximumEnergy(energy []int, k int) int {
    runningEnergy := make([]int, k)
    maxEnergy := math.MinInt
    for i := len(energy) - 1; i >= 0; i-- {
        j := i % k
        runningEnergy[j] += energy[i]
        maxEnergy = max(runningEnergy[j], maxEnergy)
    }
    return maxEnergy
}

File: 3147.taking-maximum-energy-from-the-mystic-dungeon/solution.kt
class Solution {
    fun maximumEnergy(energy: IntArray, k: Int): Int {
        val runningTotal = IntArray(k)
        var maxEnergy = Int.MIN_VALUE
        for (i in energy.lastIndex downTo 0) {
            val j = i % k
            runningTotal[j] += energy[i]
            maxEnergy = maxOf(runningTotal[j], maxEnergy)
        }
        return maxEnergy
    }
}